remote_attestation/java/src/com/google/oak/remote_attestation/Message.java

```diff
@@ -0,0 +1,356 @@
+      outputStream.flush();
+
+      return output.toByteArray();
+    }
+
+    public static ServerIdentity deserialize(byte[] input) throws IllegalArgumentException, IOException {
+      DataInputStream inputStream = new DataInputStream(new ByteArrayInputStream(input));
+
+      byte header = inputStream.readByte();
+      if (header != SERVER_IDENTITY_HEADER) {
+        throw new IllegalArgumentException("Incorrect server identity message header");
+      }
+
+      byte version = inputStream.readByte();
+      if (version != PROTOCOL_VERSION) {
+        throw new IllegalArgumentException("Incorrect remote attestation protocol version");
+      }
+
+      byte[] ephemeralPublicKey = readFixedSizeArray(inputStream, EPHEMERAL_PUBLIC_KEY_SIZE_BYTES, "ephemeral public key");
+      byte[] random = readFixedSizeArray(inputStream, REPLAY_PROTECTION_ARRAY_SIZE_BYTES, "random value");
+      byte[] transcriptSignature = readFixedSizeArray(inputStream, TRANSCRIPT_SIGNATURE_SIZE_BYTES, "transcript signature");
+      byte[] signingPublicKey = readFixedSizeArray(inputStream, SIGNING_PUBLIC_KEY_SIZE_BYTES, "signing key");
+      byte[] attestationInfo = readVariableSizeArray(inputStream, "attestation info");
+
+      return new ServerIdentity(ephemeralPublicKey, random, transcriptSignature, signingPublicKey, attestationInfo);
+    }
+  }
+
+  /**
+   * Client identity message containing remote attestation information and a public key for
+   * Diffie-Hellman key negotiation.
+   */
+  public static class ClientIdentity {
+    /** Message header. */
+    private final byte header;
+    /** Public key needed to establish a session key. */
+    private final byte[] ephemeralPublicKey;
+    /**
+     * Signature of the SHA-256 hash of all previously sent and received messages.
+     * Transcript signature is sent in messages to prevent replay attacks.
+     *
+     * Signature must be an IEEE-P1363 encoded ECDSA-P256 signature.
+     * https://datatracker.ietf.org/doc/html/rfc6979
+     * https://standards.ieee.org/standard/1363-2000.html
+     */
+    private byte[] transcriptSignature;
+    /**
+     * Public key used to sign transcripts.
+     *
+     * Public key must be an OpenSSL ECDSA-P256 key, which is represented as
+     * `0x04 | X: 32-byte | Y: 32-byte`.
+     * Where X and Y are big-endian coordinates of an Elliptic Curve point.
+     * https://datatracker.ietf.org/doc/html/rfc6979
+     */
+    private final byte[] signingPublicKey;
+    /**
+     * Information used for remote attestation such as a TEE report and a TEE provider's
+     * certificate. TEE report contains a hash of the `signingPublicKey`.
+     *
+     * Attestation info must be a serialized `oak.remote_attestation.AttestationInfo` Protobuf
+     * message.
+     */
+    private final byte[] attestationInfo;
+
+    public ClientIdentity(
+      byte[] ephemeralPublicKey,
+      byte[] transcriptSignature,
```

Comment:
Will this ever be set using the constructor to anything other than an empty byte array? If not, it is probably better to remove it from the constructor signature. Same appolies to the ServerIdentity.

Reply:
Done